helix-core/src/main/java/org/apache/helix/manager/zk/ZKHelixManager.java

```diff
@@ -780,13 +780,10 @@ public void disconnect() {
 
       // TODO reset user defined handlers only
       // TODO Fix the issue that when connection disconnected, reset handlers will be blocked. -- JJ
-      // This is because reset logic contains ZK operations.
-      resetHandlers(true);
```

Comment:
So we are not sending FINALIZE events on the reset anymore? I think this is problematic.

Reply:
Yeah. That's the design of using `unsubscribeAll`. Because the current callback handler's FINALIZE method is pretty expensive (involving read all children node under currentstates then remove them one by one). 
I could perform tests that prove without the FINALIZE event, HelixManager could clean up all information when `disconnect()`. And the single callback handler could still use the FINALIZE event to un-register itself.

Reply:
The thing is that our customers also depend on the FINALIZE event to do the cleanup. I believe with this change disconnect will clean up all the listeners. But you break the agreement.

Reply:
@jiajunwang I understand your concern. With the current helix structure & library dependency, many times there's really no easy way to satisfy everything while still keeping the code clean & short IMO, even we overall agree with the overall design here.  Let's take it offline.

I will try to resolve rest concerns as much as I could and propose a new draft code structure

Reply:
Sure, feel free to call for a short meeting with more people. My point is that we shall never trade-off correctness or functionality for performance.